### PR TITLE
Directory and file type schema

### PR DESCRIPTION
**Pull request by tetron**
This adds the `Directory` type to the schema:

* `Directory` type consists of `location`, `path` and `listing` fields
* The `listing` is a list of `Dirent` objects which specify the name and the File object of each directory entry
* Files are now identified by the `location` field.  The `path` field is now reserved for CommandLineTool context only.
* Adds `dirname`, `basename`, `nameroot` and `nameext` fields for convenience of path manipulation in CommandLineTool. 
* Allow bare string list items in `arguments` to be evaluated as expressions

### CODE REVIEW

File: draft-4/Process.yml [hunk 1]
```diff
@@ -58,28 +61,90 @@ $graph:
...
+        implementation is unable to retrieve the file content stored at a
+        remote resource (due to unsupported protocol, access denied, or other
+        issue) it must signal an error.
+      jsonldPredicate:
+        _id: "@id"
+        _type: "@id"
+    - name: path
+      type: string?
+      doc: |
+        The local path where the File is made available prior to executing a
```
Comment (mr-c) [inline draft-4/Process.yml]: This is set by the implementation, yes?

File: draft-4/conformance_test_draft-4.yaml [hunk 2]
```diff
@@ -52,54 +52,60 @@
...
-  output:
-    output_file:
-      class: File
-      size: 84
-      checksum: sha1$cec7b8746a78c42060c96505887449bca0142976
-      path: std.err
-  job:
-  tool: draft-4/egrep-stderr-mediumcut.cwl
-  doc: Test command line with stderr redirection, named brief syntax
+# - args: [egrep]
```
Comment (mr-c) [inline draft-4/conformance_test_draft-4.yaml]: Why are these tests disabled?

File: draft-4/CommandLineTool.yml [hunk 3]
```diff
@@ -626,11 +626,41 @@ $graph:
...
+        _id: "cwl:listing"
+        mapSubject: entryname
+        mapPredicate: entry
+
 
 - name: CreateFileRequirement
   type: record
   extends: "#ProcessRequirement"
   doc: |
+    *Deprecated*, superceded by InitialWorkDirRequirement.
```
Comment (mr-c) [inline draft-4/CommandLineTool.yml]: No deprecated features in 1.0, please. Lets drop it.

File: draft-4/CommandLineTool.yml [hunk 4]
```diff
@@ -626,11 +626,41 @@ $graph:
...
+    - name: entry
+      type: [string, Expression]
+      jsonldPredicate:
+        _id: cwl:entry
+
+
+- name: InitialWorkDirRequirement
+  type: record
+  extends: ProcessRequirement
+  fields:
```
Comment (mr-c) [inline draft-4/CommandLineTool.yml]: Needs documentation

**Comment by mr-c** [comment]
`stagefiles` branch of cwltool doesn't pass the tests with this branch

```
Test [66/68] hello.txt
goodbye.txt
Test failed: /home/michael/common-workflow-language/env/bin/cwltool --outdir=/tmp/tmp36YjMu --quiet draft-4/dir3.cwl draft-4/dir3-job.yml
Test directory input in Docker
expected output object {
    "outdir": {
        "class": "Directory", 
        "listing": [
            {
                "entry": {
                    "class": "File", 
                    "location": "hello.txt"
                }, 
                "entryname": "hello.txt"
            }, 
            {
                "entry": {
                    "class": "File", 
                    "location": "goodbye.txt"
                }, 
                "entryname": "goodbye.txt"
            }
        ]
    }
}
 got {
    "outdir": {
        "class": "Directory", 
        "listing": [
            {
                "entry": {
                    "class": "File", 
                    "location": "file:///tmp/tmp36YjMu/goodbye.txt", 
                    "path": "/tmp/tmp36YjMu/goodbye.txt"
                }, 
                "entryname": "goodbye.txt"
            }, 
            {
                "entry": {
                    "class": "File", 
                    "location": "file:///tmp/tmp36YjMu/hello.txt", 
                    "path": "/tmp/tmp36YjMu/hello.txt"
                }, 
                "entryname": "hello.txt"
            }
        ], 
        "location": "file:///tmp/tmp36YjMu", 
        "path": "/tmp/tmp36YjMu"
    }
}
Compare failure file:///tmp/tmp36YjMu/goodbye.txt does not end with hello.txt
Test [68/68] 1 failures, 0 unsupported features
```

**Comment by tetron** [comment]
It works for me.  Try again?

**Comment by tetron** [comment]
Oops pushed the wrong button.

**Comment by mr-c** [comment]
Same:
```
(env) michael@dan:~/common-workflow-language$ ./run_test.sh RUNNER=cwltool DRAFT=draft-4
--- Running conformance test draft-4 on /home/michael/common-workflow-language/env/bin/cwltool ---
/home/michael/common-workflow-language/env/bin/cwltool 1.0.20160629203835
Test [6/69] The file is hello.txt
Test [48/69] Hello world!
Test [66/69] hello.txt
goodbye.txt
Test failed: /home/michael/common-workflow-language/env/bin/cwltool --outdir=/tmp/tmp2hamCb --quiet draft-4/dir3.cwl draft-4/dir3-job.yml
Test directory input in Docker
expected output object {
    "outdir": {
        "class": "Directory", 
        "listing": [
            {
                "entry": {
                    "class": "File", 
                    "location": "hello.txt"
                }, 
                "entryname": "hello.txt"
            }, 
            {
                "entry": {
                    "class": "File", 
                    "location": "goodbye.txt"
                }, 
            }
        ]
    }
}
 got {
    "outdir": {
        "class": "Directory", 
        "listing": [
            {
                "entry": {
                    "class": "File", 
                    "location": "file:///tmp/tmp2hamCb/goodbye.txt", 
                    "path": "/tmp/tmp2hamCb/goodbye.txt"
                }, 
                "entryname": "goodbye.txt"
            }, 
            {
                "entry": {
                    "class": "File", 
                    "location": "file:///tmp/tmp2hamCb/hello.txt", 
                    "path": "/tmp/tmp2hamCb/hello.txt"
                }, 
                "entryname": "hello.txt"
            }
        ], 
        "location": "file:///tmp/tmp2hamCb", 
        "path": "/tmp/tmp2hamCb"
    }
}
Compare failure file:///tmp/tmp2hamCb/goodbye.txt does not end with hello.txt
Test [69/69] 1 failures, 0 unsupported features

1 tool tests failed
```
common-workflow-language is at cae0407
cwltool is at d2270a7

**Comment by mr-c** [comment]
(tested on two systems)

**Comment by tetron** [comment]
Oh I think I see now. The order is not consistent.  I should probably teach cwltest  that "listing" doesn't need to be exact order.

**Comment by tetron** [comment]
This should work with the latest stagefiles branch

**Comment by mr-c** [comment]
Hmm..

```
(env) michael@dan:~/common-workflow-language$ ./run_test.sh RUNNER=cwltool DRAFT=draft-4                                                               
--- Running conformance test draft-4 on /home/michael/common-workflow-language/env/bin/cwltool ---
/home/michael/common-workflow-language/env/bin/cwltool 1.0.20160630145357
Test [6/69] The file is hello.txt
Test failed: /home/michael/common-workflow-language/env/bin/cwltool --outdir=/tmp/tmpWEbeVM --quiet draft-4/template-tool.cwl draft-4/cat-job.json
Test CreateFileRequirement ExpressionEngineRequirement.engineConfig feature
expected output object {
    "foo": {
        "checksum": "sha1$63da67422622fbf9251a046d7a34b7ea0fd4fead", 
        "class": "File", 
        "location": "foo.txt", 
        "size": 22
    }
}
 got {
    "foo": {
        "checksum": "sha1$63da67422622fbf9251a046d7a34b7ea0fd4fead", 
        "class": "File", 
        "location": "file:///tmp/tmpWEbeVM/foo.txt", 
        "path": "/tmp/tmpWEbeVM/foo.txt", 
        "size": 22
    }
}
```

**Pull request merged**